### PR TITLE
ORM auto-alias

### PR DESCRIPTION
**Pull request by steefaan**
refs https://github.com/cakephp/cakephp/issues/7050

Before the changes you need to do:

```
$columns = $this->schema()->columns();
$subQuery = $this->find();
$subQuery->select(array_combine($columns, $columns));
$query = $this->find()->from(['Geolocations' => $subQuery]);
```

With the changes you can do:

```
$subQuery = $this->find()->autoAlias(false);
$query = $this->find()->from(['Geolocations' => $subQuery]);
```

If you are fine with the changes, I will do the test cases.

### CODE REVIEW

File: src/Database/QueryCompiler.php [hunk 1]
```diff
@@ -153,7 +153,7 @@ protected function _buildSelectPart($parts, $query, $generator)
         $normalized = [];
         $parts = $this->_stringifyExpressions($parts, $generator);
         foreach ($parts as $k => $p) {
-            if (!is_numeric($k)) {
+            if ($query->autoAlias() && !is_numeric($k)) {
```
Comment (steefaan) [inline src/Database/QueryCompiler.php]: As feared not the best solution here (see travis). I will try to find a better solution. Probably it's better to handle in `ORM/Query.php`.

File: src/ORM/Query.php [hunk 2]
```diff
@@ -367,6 +375,10 @@ public function aliasField($field, $alias = null)
             $aliasedField = $alias . '.' . $field;
         }
 
+        if ($this->_autoAlias === false) {
+            return [$aliasedField];
```
Comment (steefaan) [inline src/ORM/Query.php]: Try to build a numeric array to react on https://github.com/cakephp/cakephp/blob/master/src/Database/QueryCompiler.php#L156

Unfortunately there is another [potential problem](https://github.com/cakephp/cakephp/blob/master/src/ORM/ResultSet.php#L519) when numeric array needs to be handled. The `array_intersect_key` isn't able to work properly with `->first()`. Bring an extra condition to `ResultSet` feels a bit messy. Someone has another idea?

**Comment by markstory** [comment]
Mode operations require pretty extensive test coverage as they add effect all the associations, custom fields and possibly join condition generation. I know thst hydrate() has a ton of tests around it for the same reasons.

**Comment by lorenzo** [comment]
@steefaan would you add some tests to this?

**Comment by steefaan** [comment]
AS @markstory mentoined it requires pretty extensive test coverage.If I remember correctly this works fine for find('all'); but I had issues with find()->first() and I'm not sure if I'm able to solve these issues. In general I'm not sure if the suggested approach is the best way overall.

**Comment by markstory** [comment]
Closing as this cannot be merged, and I'm a little bit concerned about the complexity that more modal states introduce to the query builder.

**Pull request closed without merging**